File: sdkwork-rtc-sdk-csharp/src/SDKWork.Rtc.Sdk/RtcDriverManager.cs
namespace Sdkwork.Rtc.Sdk;

using System.Linq;

public sealed class RtcDriverManager
{
    public RtcProviderSelection ResolveSelection(
        RtcProviderSelectionRequest? request = null,
        string defaultProviderKey = RtcProviderCatalog.DEFAULT_RTC_PROVIDER_KEY
    )
    {
        return RtcProviderSelection.ResolveRtcProviderSelection(request, defaultProviderKey);
    }

    public RtcProviderSupport DescribeProviderSupport(string providerKey)
    {
        var official = RtcProviderCatalog.GetRtcProviderByProviderKey(providerKey) is not null;
        var activation = RtcProviderActivationCatalog.GetRtcProviderActivationByProviderKey(providerKey);

        return RtcProviderSupport.CreateRtcProviderSupportState(
            new RtcProviderSupportStateRequest(
                providerKey,
                activation?.builtin ?? false,
                official,
                activation?.runtimeBridge ?? false
            )
        );
    }

    public IReadOnlyList<RtcProviderSupport> ListProviderSupport()
    {
        return RtcProviderCatalog.Entries
            .Select(entry => DescribeProviderSupport(entry.providerKey))
            .ToArray();
    }
}

File: sdkwork-rtc-sdk-swift/Sources/RtcSdk/RtcDriverManager.swift
public struct RtcDriverManager {
    public init() {}

    public func resolveSelection(
        request: RtcProviderSelectionRequest = RtcProviderSelectionRequest(),
        defaultProviderKey: String = RtcProviderCatalog.DEFAULT_RTC_PROVIDER_KEY
    ) -> RtcProviderSelection {
        return RtcProviderSelection.resolveRtcProviderSelection(
            request: request,
            defaultProviderKey: defaultProviderKey
        )
    }

    public func describeProviderSupport(providerKey: String) -> RtcProviderSupport {
        let official = RtcProviderCatalog.getRtcProviderByProviderKey(providerKey) != nil
        let activation = RtcProviderActivationCatalog.getRtcProviderActivationByProviderKey(providerKey)

        return RtcProviderSupport.createRtcProviderSupportState(
            .init(
                providerKey: providerKey,
                builtin: activation?.builtin ?? false,
                official: official,
                registered: activation?.runtimeBridge ?? false
            )
        )
    }

    public func listProviderSupport() -> [RtcProviderSupport] {
        RtcProviderCatalog.entries.map { describeProviderSupport(providerKey: $0.providerKey) }
    }
}
